New: Add JSLL script check

qzhou1607 opened the pull request:
New Rule to check if JSLL script is in the page head

<!--

Read our pull request guide:
https://sonarwhal.com/docs/contributor-guide/contributing/pull-requests/

For the following items put an "x" between the square brackets
(i.e. [x]) if you completed the associated item.

-->

## Pull request checklist

Make sure you:

- [X] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [X] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/contributing/pull-requests/#commitmessages)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [X] Added/Updated related tests.

## Short description of the change(s)
* Validate that JSLL script is in `<head>` tag.
* Validate that the JSLL script is placed prior to all other scripts.
* Verify version number format in the file names.

<!--

If this is a non-trivial change, include information such as what
benefits this change brings as well as possible drawbacks.

If this fixes an existing issue, include the relavant issue number(s).

Thank you for taking the time to open this PR!

-->

molant commented:
Adding the `blocked` label because this should go into another package as soon as #685 is finished and merged

molant commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+            }
+
+            const pageDOM: IAsyncHTMLDocument = context.pageDOM as IAsyncHTMLDocument;
+            const scriptsInHead: Array<IAsyncHTMLElement> = await pageDOM.querySelectorAll('head script');
```
We should be using the `element::script` event instead of `traverse::end`.
This should make things slightly faster in some connectors because you will not have to query the DOM to get the elements and you make a better of use of our API.

qzhou1607 replied:
@molant Hmm... How do you verify that the script is in `head` then?

molant replied:
Vía the traverse::up/down events. If you need more information in those events to make it easier let me know.

qzhou1607 replied:
@molant I split the `validate` function into two validate functions, and attach them to `element::head` and `element::script` separately. By listening to the `element::head` event,  it checks the number/location of the jsll scripts without throwing duplicate messages. Does this work?

molant replied:
You are still doing `querySelectorAll` at the page level. If you prefer to continue down this path then do it from the element (do we support `querySelectorAll` at the element level?). If not it is just the same thing as before but being executed in another event.

What I'd do is something like:

1. subscribe to head event or `traverse::down` to the heard
1. set a property `inHead = true`
1. in the `element::script` event, check if you are in the head via the `inHead` property. If not, quit
1. do all the validations you need to do there
1. when you do the `traverse::up`, set `inHead` to `false`

something like:

```js
let inHead = false;

const traverse = (data) =>{
  if( traverseDown of head) {
    inHead = true;
    return;
  }
 
  if (traverseUp of head) {
    inHead = false;
    return;
  }
};

const validateScript = (data) => {
  // validations here taking into account the value of inHead
};
```

Does it make sense?

molant commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+            category: Category.performance,
+            description: `This rule confirms that JSLL script is included in the head of the page`
+        },
+        recommended: true,
```
should be false

molant commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+    },
+    meta: {
+        docs: {
+            category: Category.performance,
```
We probably need to create another category. @sarvaje, @alrra, do we have another category for custom rules? Should we create one? Or should the category be empty for custom rules?

molant commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+        const validate = async (data: ITraverseEnd) => {
+            const { resource }: { resource: string } = data;
+            const jsllDir = `https://az725175.vo.msecnd.net/scripts/jsll-`;
+            const passRegex = new RegExp(`^(\\d+\\.)js`);
```
Add an example of a version that passes

molant commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+            const { resource }: { resource: string } = data;
+            const jsllDir = `https://az725175.vo.msecnd.net/scripts/jsll-`;
+            const passRegex = new RegExp(`^(\\d+\\.)js`);
+            const warningRegex = new RegExp(`^(\\d+\\.){2,}js`);
```
Add an example of version that fails

alrra commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+            }
+
+            if (jsllScriptsInHead.length > 1) {
+                await context.report(resource, null, redundantScriptInHeadMsg);
```
Maybe flag each script that shouldn't be included? That way it's easier to know exactly which scripts should be removed.

alrra commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+            const invalidScriptVersionMsg = `The jsll script versioning is not valid.`;
+            const wrongScriptOrderMsg = `The JSLL script isn't placed prior to other scripts.`;
+
+            const getJsllScripts= (elements: Array<IAsyncHTMLElement>) => {
```
Maybe this should be outside of `validate`?

alrra commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+            const jsllScript: IAsyncHTMLElement = jsllScriptsInHead.pop();
+
+            if (scriptsInHead.indexOf(jsllScript) !== 0) {
+                await context.report(resource, null, wrongScriptOrderMsg);
```
Replace `null` with the actual script element.

alrra commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+                return;
+            }
+
+            const fileName: string = jsllScript.getAttribute('src').replace(jsllDir, '');
```
`normalizeString(jsllScript.getAttribute('src'))`

alrra commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+            }
+
+            if (warningRegex.test(fileName)) {
+                await context.report(resource, null, warningScriptVersionMsg, null, null, Severity.warning);
```
Replace `null` with the actual script element.

alrra commented on src/lib/rules/jsll-script-included/jsll-script-included.ts:
```diff
@@ -0,0 +1,94 @@
+                return;
+            }
+
+            await context.report(resource, null, invalidScriptVersionMsg);
```
Replace `null` with the actual script element.

qzhou1607 commented:
@molant  I added additional `element` information in the `traverse` events and changed the code accordingly. Can you review the PR again? I will squash the commits once it's approved.

molant commented:
This PR is being handled in the repo for Microsoft rules so closing.
